fix on_recv called when None after close

graingert opened the pull request:
Fixes #82

graingert committed:
add test for ipc Connection.close()

graingert committed:
fix on_recv called when None after close

graingert commented on pytest.ini:
```diff
@@ -2,4 +2,5 @@
-    tests
\ No newline at end of file
+    tests
+filterwarnings = error
```
this is so pytest will fail tests that issue https://docs.pytest.org/en/7.1.x/reference/reference.html#pytest.PytestUnhandledThreadExceptionWarning

graingert commented:
@mmerickel can you approve my workflow build please?

graingert commented:
thanks looks like there's an fd use after close bug:

```
>               warnings.warn(pytest.PytestUnhandledThreadExceptionWarning(msg))
E               pytest.PytestUnhandledThreadExceptionWarning: Exception in thread Thread-1
E               
E               Traceback (most recent call last):
E                 File "/Users/runner/hostedtoolcache/PyPy/3.8.16/x64/lib/pypy3.8/threading.py", line 932, in _bootstrap_inner
E                   self.run()
E                 File "/Users/runner/hostedtoolcache/PyPy/3.8.16/x64/lib/pypy3.8/threading.py", line 870, in run
E                   self._target(*self._args, **self._kwargs)
E                 File "/Users/runner/work/hupper/hupper/.tox/py/lib/pypy3.8/site-packages/hupper/ipc.py", line 167, in _read_loop
E                   packet = self._recv_packet()
E                 File "/Users/runner/work/hupper/hupper/.tox/py/lib/pypy3.8/site-packages/hupper/ipc.py", line 144, in _recv_packet
E                   chunk = os.read(self.r_fd, self._packet_len.size)
E               OSError: [Errno 9] Bad file descriptor
```

graingert committed:
suppress EBADF errors and set ipc fds to -1 on close to avoid use after close

graingert commented:
@mmerickel can I get another run approved please?

graingert commented on src/hupper/ipc.py:
```diff
@@ -135,8 +136,11 @@ def activate(self, on_recv):
-        close_fd(self.r_fd)
-        close_fd(self.w_fd)
+        self.r_fd, r_fd = -1, self.r_fd
+        self.w_fd, w_fd = -1, self.r_fd
```
```suggestion
        self.w_fd, w_fd = -1, self.w_fd
```

graingert committed:
fix typo in ipc.close()

mmerickel commented on src/hupper/ipc.py:
```diff
@@ -157,16 +161,23 @@ def _recv_packet(self):
+        except OSError as e:
+            if e.errno != errno.EBADF:
+                raise
+        self._on_recv(None)
```
what is the goal here of calling `_on_recv(None)`?

graingert replied:
I don't know, the code was doing it before on line 169

mmerickel commented on src/hupper/ipc.py:
```diff
@@ -135,8 +136,11 @@ def activate(self, on_recv):
+
+        close_fd(r_fd)
+        close_fd(w_fd)
         self.on_recv = None
```
I'm trying to remember why the reader thread wouldn't be joined on here just to ensure proper cleanup.

mmerickel replied:
To be clear, wouldn't a `reader_thread.join()` prior to `self.on_recv = None` fix the issue without needing to overwrite the fd values?

graingert replied:
No, the thread could be executing on_recv when the FD is closed, meaning when it goes around the loop again it tries to read from a stale FD value

graingert replied:
you also can't join the reader thread because it might not have been created, eg the child_pipe of worker doesn't get "activate()" called on it

graingert commented:
~Apparently os.read(-1) hangs on windows?~

looks like Connection.close() hangs on windows, this happens on main too if Connection is tested in isolation

graingert commented:
this always hangs forever on windows:

```python
import os
import sys
import threading

def consume(ident, fd):
    while chunk := os.read(fd, 1):
        print(ident, chunk)

def main():
    cr, cw = os.pipe()
    crt = threading.Thread(target=consume, args=("cr", cr))
    crt.start()

    print("close cr", flush=True)
    os.close(cr)
    print("close cw", flush=True)
    os.close(cw)

    

if __name__ == "__main__":
    sys.exit(main())
```

graingert committed:
only activate one end of pipe per process

windows serializes all access to pipes so a close will block until a
read returns
https://groups.google.com/g/microsoft.public.win32.programmer.kernel/c/VzaZ41mDcVs/m/zSshHYVXyvQJ

graingert commented:
@mmerickel ok I tested this on linux and windows now it should pass in CI!

shughes-uk commented:
This is a constant annoyance I have using hupper, would love to see it fixed!

mmerickel commented:
@graingert I finally got a chance to sit down and think about this and I'm wondering if we can simplify this to a one-line fix? What do you think?

```diff
diff --git a/src/hupper/ipc.py b/src/hupper/ipc.py
index cec1a85..11fdba3 100644
--- a/src/hupper/ipc.py
+++ b/src/hupper/ipc.py
@@ -135,9 +135,9 @@ class Connection(object):
         self.reader_thread.start()
 
     def close(self):
+        self.on_recv = lambda _: None
         close_fd(self.r_fd)
         close_fd(self.w_fd)
-        self.on_recv = None
 
     def _recv_packet(self):
         buf = io.BytesIO()
```

graingert commented:
Doesn't fix the use after close

graingert committed:
assign on_recv to noop lambda on ipc.Connection.close

graingert commented:
@mmerickel I've pushed that change

mmerickel commented:
How does it not fix the use-after-close? The exception you posted in the original ticket is showing that things were properly caught (by the EOFError?) but the on_recv was None at the time instead of something that could be invoked.

mmerickel commented:
ok I see - it's not in your original stacktrace but I agree that invoking read on a closed fd will return EBADF which is not currently handled so your fix improves that.

mmerickel commented:
Please modify the PR to set self.on_recv before closing the fds to avoid a race there. I think there is no real point in setting the fds to -1 either but if you want to keep that it's ok. Thanks!

The following to me is a proper fix:

```diff
diff --git a/src/hupper/ipc.py b/src/hupper/ipc.py
index cec1a85..bbc567a 100644
--- a/src/hupper/ipc.py
+++ b/src/hupper/ipc.py
@@ -1,3 +1,4 @@
+import errno
 import io
 import os
 import pickle
@@ -135,9 +136,9 @@ class Connection(object):
         self.reader_thread.start()
 
     def close(self):
+        self.on_recv = lambda _: None
         close_fd(self.r_fd)
         close_fd(self.w_fd)
-        self.on_recv = None
 
     def _recv_packet(self):
         buf = io.BytesIO()
@@ -166,6 +167,9 @@ class Connection(object):
                 self.on_recv(packet)
         except EOFError:
             pass
+        except OSError as e:
+            if e.errno != errno.EBADF:
+                raise
         self.on_recv(None)
 
     def _write_packet(self, data):
```

mmerickel commented:
@graingert I haven't had time to debug this but something is clearly wrong with the new test you added on windows. Obviously one run passes, but all the others fail. I tried to re-run the failed jobs and they failed again in the same way with a deadlock in those tests.

mmerickel committed:
close on_recv first, before closing fds

mmerickel committed:
close the writer before the reader to fix a blocking issue on windows

You cannot invoke close(r_fd) on windows while a blocking os.read(r_fd)
is active. However, if you close the write_fd first, then things seem to
work.

mmerickel commented:
Whelp, 4 hours later after installing a windows VM and banging my head for a while I found the issue. Basically you cannot invoke `os.close(r_fd)` if the other thread is blocking on `os.read(r_fd)`. Thanks windows. It appears to work reliably if you close the write_fd before the read_fd.
